tsfc/loopy.py

```diff
@@ -0,0 +1,340 @@
+"""Generate loopy kernel from ImperoC tuple data.
+
+This is the final stage of code generation in TSFC."""
+
+from __future__ import absolute_import, print_function, division
+
+from math import isnan
+
+import numpy
+from functools import singledispatch
+from collections import defaultdict, OrderedDict
+
+from gem import gem, impero as imp
+
+import islpy as isl
+import loopy as lp
+
+import pymbolic.primitives as p
+
+from pytools import UniqueNameGenerator
+
+
+class LoopyContext(object):
+    def __init__(self):
+        self.indices = {}
+        self.active_indices = {}  # gem index -> pymbolic variable
+        self.index_extent = OrderedDict()  # pymbolic variable for indices - > extent
+        self.gem_to_pymbolic = {}  # gem node -> pymbolic variable
+        self.name_gen = UniqueNameGenerator()
+
+    def pym_multiindex(self, multiindex):
+        rank = []
+        for index in multiindex:
+            if isinstance(index, gem.Index):
+                rank.append(self.active_indices[index])
+            elif isinstance(index, gem.VariableIndex):
+                rank.append(expression(index.expression, self))
+            else:
+                assert isinstance(index, int)
+                rank.append(index)
+        return tuple(rank)
```

Comment:
rank is definitely the wrong noun, why not "indices"?

Reply:
Amended.